core/trino-spi/src/main/java/io/trino/spi/type/Int128Math.java

```diff
@@ -0,0 +1,1341 @@
+        shiftRightMultiPrecision(dividend, normalizedDividendLength, nlz);
+    }
+
+    private static void divideKnuthNormalized(int[] remainder, int dividendLength, int[] divisor, int divisorLength, int[] quotient)
+    {
+        int v1 = divisor[divisorLength - 1];
+        int v0 = divisor[divisorLength - 2];
+        for (int reminderIndex = dividendLength - 1; reminderIndex >= divisorLength; reminderIndex--) {
+            int qHat = estimateQuotient(remainder[reminderIndex], remainder[reminderIndex - 1], remainder[reminderIndex - 2], v1, v0);
+            if (qHat != 0) {
+                boolean overflow = multiplyAndSubtractUnsignedMultiPrecision(remainder, reminderIndex, divisor, divisorLength, qHat);
+                // Add back - probability is 2**(-31). R += D. Q[digit] -= 1
+                if (overflow) {
+                    qHat--;
+                    addUnsignedMultiPrecision(remainder, reminderIndex, divisor, divisorLength);
+                }
+            }
+            quotient[reminderIndex - divisorLength] = qHat;
+        }
+    }
+
+    /**
+     * Use the Knuth notation
+     * <p>
+     * u{x} - dividend
+     * v{v} - divisor
+     */
+    private static int estimateQuotient(int u2, int u1, int u0, int v1, int v0)
+    {
+        // estimate qhat based on the first 2 digits of divisor divided by the first digit of a dividend
+        long u21 = combineInts(u2, u1);
+        long qhat;
+        if (u2 == v1) {
+            qhat = INT_BASE - 1;
+        }
+        else if (u21 >= 0) {
+            qhat = u21 / toUnsignedLong(v1);
+        }
+        else {
+            qhat = divideUnsignedLong(u21, v1);
+        }
+
+        if (qhat == 0) {
+            return 0;
+        }
+
+        // Check if qhat is greater than expected considering only first 3 digits of a dividend
+        // This step help to eliminate all the cases when the estimation is greater than q by 2
+        // and eliminates most of the cases when qhat is greater than q by 1
+        //
+        // u2 * b * b + u1 * b + u0 >= (v1 * b + v0) * qhat
+        // u2 * b * b + u1 * b + u0 >= v1 * b * qhat + v0 * qhat
+        // u2 * b * b + u1 * b - v1 * b * qhat >=  v0 * qhat - u0
+        // (u21 - v1 * qhat) * b >=  v0 * qhat - u0
+        // (u21 - v1 * qhat) * b + u0 >=  v0 * qhat
+        // When ((u21 - v1 * qhat) * b + u0) is less than (v0 * qhat) decrease qhat by one
+
+        int iterations = 0;
+        long rhat = u21 - toUnsignedLong(v1) * qhat;
+        while (Long.compareUnsigned(rhat, INT_BASE) < 0 && Long.compareUnsigned(toUnsignedLong(v0) * qhat, combineInts(lowInt(rhat), u0)) > 0) {
+            iterations++;
+            qhat--;
+            rhat += toUnsignedLong(v1);
+        }
+
+        if (iterations > 2) {
+            throw new IllegalStateException("qhat is greater than q by more than 2: " + iterations);
+        }
+
+        return (int) qhat;
+    }
+
+    private static long divideUnsignedLong(long dividend, int divisor)
```

Comment:
Can this be `Long.divideUnsigned()`

Reply:
Yes, but it's pre-existent. I'll change it once I get the new division code in.